File: src/gear/sandboxing.py
import os
import subprocess
import shlex
import shutil
from enum import Enum
from typing import List, Dict, Optional, Tuple, Callable
from dataclasses import dataclass, field
from pathlib import Path


class IsolationLevel(Enum):
    STRICT = "strict"
    MEDIUM = "medium"
    MINIMAL = "minimal"


class SandboxStatus(Enum):
    READY = "ready"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    TIMEOUT = "timeout"


@dataclass
class SandboxOptions:
    isolation_level: IsolationLevel = IsolationLevel.MEDIUM
    allow_network: bool = False
    allow_home: bool = False
    allow_x11: bool = False
    allow_dbus: bool = False
    working_dir: Optional[str] = None
    bind_paths: List[Tuple[str, str, Optional[str]]] = field(default_factory=list)
    env_vars: Dict[str, str] = field(default_factory=dict)
    capabilities: List[str] = field(default_factory=list)
    uid: Optional[int] = None
    gid: Optional[int] = None
    hostname: Optional[str] = None
    tmpfs_paths: List[str] = field(default_factory=list)
    ro_bind_paths: List[Tuple[str, str]] = field(default_factory=list)
    dev_bind_paths: List[Tuple[str, str]] = field(default_factory=list)


@dataclass
class SandboxResult:
    return_code: int
    stdout: str
    stderr: str
    status: SandboxStatus
    execution_time: float = 0.0
    warnings: List[str] = field(default_factory=list)
    errors: List[str] = field(default_factory=list)


class SandboxManager:
    
    def __init__(self):
        self._check_dependencies()
        self._running_processes: Dict[str, subprocess.Popen] = {}
        
    def _check_dependencies(self):
        if not self._is_bwrap_available():
            raise RuntimeError("bubblewrap (bwrap) is required but not found")
        
        if not self._verify_bwrap_functionality():
            raise RuntimeError("bubblewrap (bwrap) is not functioning correctly")
    
    def _is_bwrap_available(self) -> bool:
        return shutil.which('bwrap') is not None
    
    def _verify_bwrap_functionality(self) -> bool:
        try:
            result = subprocess.run(
                ['bwrap', '--version'],
                capture_output=True,
                text=True,
                timeout=5
            )
            return result.returncode == 0
        except:
            return False
    
    def create_sandbox_options(self, isolation_level: IsolationLevel = IsolationLevel.MEDIUM,
                             **kwargs) -> SandboxOptions:
        options = SandboxOptions(isolation_level=isolation_level)
        
        for key, value in kwargs.items():
            if hasattr(options, key):
                setattr(options, key, value)
        
        return options
    
    def execute_sandboxed_makepkg(self, build_dir: str,
                                 makepkg_args: Optional[List[str]] = None,
                                 options: Optional[SandboxOptions] = None) -> bool:
        
        if not options:
            options = SandboxOptions(
                isolation_level=IsolationLevel.MEDIUM,
                allow_network=True,
                working_dir=build_dir
            )
        
        makepkg_cmd = ['makepkg']
        if makepkg_args:
            makepkg_cmd.extend(makepkg_args)
        else:
            makepkg_cmd.extend(['-s', '-r', '-c'])
        
        result = self.run_sandboxed_command(makepkg_cmd, options)
        return result.status == SandboxStatus.COMPLETED and result.return_code == 0
    
    def execute_sandboxed_paru(self, paru_args: List[str],
                              options: Optional[SandboxOptions] = None) -> SandboxResult:
        
        if not options:
            options = SandboxOptions(
                isolation_level=IsolationLevel.MEDIUM,
                allow_network=True,
                allow_home=True
            )
        
        paru_cmd = ['paru'] + paru_args
        return self.run_sandboxed_command(paru_cmd, options)
    
    def run_sandboxed_command(self, command: List[str],
                             options: SandboxOptions,
                             timeout: Optional[int] = None,
                             output_callback: Optional[Callable[[str, str], None]] = None) -> SandboxResult:
        
        import time
        start_time = time.time()
        
        try:
            bwrap_args = self._build_bwrap_command(command, options)
            
            process = subprocess.Popen(
                bwrap_args,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                bufsize=1,
                universal_newlines=True,
                cwd=options.working_dir
            )
            
            stdout_lines = []
            stderr_lines = []
            
            try:
                while True:
                    stdout_line = process.stdout.readline()
                    stderr_line = process.stderr.readline()
                    
                    if stdout_line:
                        stdout_lines.append(stdout_line)
                        if output_callback:
                            output_callback(stdout_line.rstrip(), "stdout")
                    
                    if stderr_line:
                        stderr_lines.append(stderr_line)
                        if output_callback:
                            output_callback(stderr_line.rstrip(), "stderr")
                    
                    if not stdout_line and not stderr_line and process.poll() is not None:
                        break
                
                remaining_stdout, remaining_stderr = process.communicate(timeout=timeout)
                if remaining_stdout:
                    stdout_lines.extend(remaining_stdout.splitlines(keepends=True))
                if remaining_stderr:
                    stderr_lines.extend(remaining_stderr.splitlines(keepends=True))
                
                execution_time = time.time() - start_time
                
                stdout_str = ''.join(stdout_lines)
                stderr_str = ''.join(stderr_lines)
                
                result = SandboxResult(
                    return_code=process.returncode,
                    stdout=stdout_str,
                    stderr=stderr_str,
                    status=SandboxStatus.COMPLETED if process.returncode == 0 else SandboxStatus.FAILED,
                    execution_time=execution_time
                )
                
                result.warnings = self._extract_warnings(stdout_str, stderr_str)
                result.errors = self._extract_errors(stderr_str)
                
                return result
                
            except subprocess.TimeoutExpired:
                process.kill()
                process.wait()
                
                return SandboxResult(
                    return_code=-1,
                    stdout=''.join(stdout_lines),
                    stderr=''.join(stderr_lines),
                    status=SandboxStatus.TIMEOUT,
                    execution_time=time.time() - start_time,
                    errors=["Command timed out"]
                )
                
        except Exception as e:
            return SandboxResult(
                return_code=-1,
                stdout="",
                stderr=str(e),
                status=SandboxStatus.FAILED,
                execution_time=time.time() - start_time,
                errors=[f"Execution error: {str(e)}"]
            )
    
    def _build_bwrap_command(self, command: List[str], options: SandboxOptions) -> List[str]:
        args = ['bwrap']
        
        args.extend(['--unshare-all', '--die-with-parent'])
        
        if options.isolation_level == IsolationLevel.STRICT:
            args.extend(['--unshare-user', '--unshare-ipc', '--unshare-pid', '--unshare-cgroup'])
            args.extend(['--no-new-privileges'])
            args.extend(['--cap-drop', 'ALL'])
            
            for cap in options.capabilities:
                args.extend(['--cap-add', cap])
        
        elif options.isolation_level == IsolationLevel.MEDIUM:
            args.extend(['--unshare-user', '--unshare-ipc'])
            
        elif options.isolation_level == IsolationLevel.MINIMAL:
            pass
        
        args.extend(['--dev', '/dev'])
        args.extend(['--proc', '/proc'])
        args.extend(['--tmpfs', '/tmp'])
        args.extend(['--tmpfs', '/var/tmp'])
        
        for tmpfs_path in options.tmpfs_paths:
            args.extend(['--tmpfs', tmpfs_path])
        
        essential_ro_binds = [
            '/usr', '/bin', '/lib', '/lib64', '/sbin', '/etc'
        ]
        
        for path in essential_ro_binds:
            if os.path.exists(path):
                args.extend(['--ro-bind', path, path])
        
        for host_path, container_path in options.ro_bind_paths:
            if os.path.exists(host_path):
                args.extend(['--ro-bind', host_path, container_path])
        
        for host_path, container_path in options.dev_bind_paths:
            if os.path.exists(host_path):
                args.extend(['--dev-bind', host_path, container_path])
        
        for host_path, container_path, flags in options.bind_paths:
            if os.path.exists(host_path):
                bind_type = flags if flags else '--bind'
                args.extend([bind_type, host_path, container_path])
        
        if options.allow_network:
            args.extend(['--share-net'])
        else:
            args.extend(['--unshare-net'])
        
        if options.allow_home:
            home_dir = os.path.expanduser('~')
            args.extend(['--bind', home_dir, home_dir])
        else:
            args.extend(['--tmpfs', '/home'])
            fake_home = '/home/sandbox'
            args.extend(['--mkdir', fake_home])
            args.extend(['--setenv', 'HOME', fake_home])
        
        if options.allow_x11:
            x11_socket = os.environ.get('DISPLAY')
            if x11_socket and os.path.exists('/tmp/.X11-unix'):
                args.extend(['--bind', '/tmp/.X11-unix', '/tmp/.X11-unix'])
                args.extend(['--setenv', 'DISPLAY', x11_socket])
        
        if options.allow_dbus:
            dbus_session = os.environ.get('DBUS_SESSION_BUS_ADDRESS')
            if dbus_session:
                args.extend(['--setenv', 'DBUS_SESSION_BUS_ADDRESS', dbus_session])
                
                xdg_runtime_dir = os.environ.get('XDG_RUNTIME_DIR')
                if xdg_runtime_dir and os.path.exists(xdg_runtime_dir):
                    args.extend(['--bind', xdg_runtime_dir, xdg_runtime_dir])
        
        if options.working_dir:
            if not os.path.exists(options.working_dir):
                try:
                    os.makedirs(options.working_dir, exist_ok=True)
                except:
                    pass
            
            args.extend(['--bind', options.working_dir, options.working_dir])
            args.extend(['--chdir', options.working_dir])
        else:
            args.extend(['--chdir', '/tmp'])
        
        if options.uid is not None:
            args.extend(['--uid', str(options.uid)])
        
        if options.gid is not None:
            args.extend(['--gid', str(options.gid)])
        
        if options.hostname:
            args.extend(['--hostname', options.hostname])
        
        default_env = {
            'PATH': '/usr/bin:/bin:/usr/sbin:/sbin',
            'LANG': os.environ.get('LANG', 'C.UTF-8'),
            'USER': 'sandbox',
            'LOGNAME': 'sandbox'
        }
        
        for key, value in default_env.items():
            args.extend(['--setenv', key, value])
        
        for key, value in options.env_vars.items():
            args.extend(['--setenv', key, value])
        
        pacman_paths = [
            '/var/lib/pacman',
            '/var/cache/pacman',
            '/etc/pacman.conf',
            '/etc/pacman.d'
        ]
        
        for path in pacman_paths:
            if os.path.exists(path):
                args.extend(['--ro-bind', path, path])
        
        args.append('--')
        args.extend(command)
        
        return args
    
    def _extract_warnings(self, stdout: str, stderr: str) -> List[str]:
        warnings = []
        combined_output = stdout + stderr
        
        warning_patterns = [
            'warning:',
            'caution:',
            'deprecated',
            'bwrap:',
            'permission denied',
            'access denied'
        ]
        
        for line in combined_output.splitlines():
            line_lower = line.lower()
            for pattern in warning_patterns:
                if pattern in line_lower:
                    warnings.append(line.strip())
                    break
        
        return warnings
    
    def _extract_errors(self, stderr: str) -> List[str]:
        errors = []
        
        error_patterns = [
            'error:',
            'failed:',
            'cannot',
            'unable to',
            'not found',
            'permission denied',
            'access denied'
        ]
        
        for line in stderr.splitlines():
            line_lower = line.lower()
            for pattern in error_patterns:
                if pattern in line_lower:
                    errors.append(line.strip())
                    break
        
        return errors
    
    def test_sandbox_functionality(self) -> Tuple[bool, List[str]]:
        test_results = []
        all_passed = True
        
        try:
            options = SandboxOptions(
                isolation_level=IsolationLevel.MINIMAL,
                allow_network=False,
                allow_home=False
            )
            
            test_cmd = ['echo', 'sandbox test']
            result = self.run_sandboxed_command(test_cmd, options, timeout=10)
            
            if result.status == SandboxStatus.COMPLETED and result.return_code == 0:
                test_results.append("Basic sandbox execution: PASS")
            else:
                test_results.append("Basic sandbox execution: FAIL")
                all_passed = False
                
        except Exception as e:
            test_results.append(f"Basic sandbox execution: FAIL - {str(e)}")
            all_passed = False
        
        try:
            options = SandboxOptions(
                isolation_level=IsolationLevel.STRICT,
                allow_network=False,
                allow_home=False
            )
            
            network_test_cmd = ['ping', '-c', '1', '8.8.8.8']
            result = self.run_sandboxed_command(network_test_cmd, options, timeout=10)
            
            if result.return_code != 0:
                test_results.append("Network isolation test: PASS")
            else:
                test_results.append("Network isolation test: FAIL - Network access not blocked")
                all_passed = False
                
        except Exception as e:
            test_results.append(f"Network isolation test: FAIL - {str(e)}")
            all_passed = False
        
        return all_passed, test_results
    
    def get_sandbox_info(self) -> Dict[str, any]:
        info = {
            'bwrap_available': self._is_bwrap_available(),
            'bwrap_functional': False,
            'bwrap_version': 'unknown',
            'running_processes': len(self._running_processes),
            'supported_features': []
        }
        
        if info['bwrap_available']:
            try:
                result = subprocess.run(
                    ['bwrap', '--version'],
                    capture_output=True,
                    text=True,
                    timeout=5
                )
                
                if result.returncode == 0:
                    info['bwrap_functional'] = True
                    version_line = result.stdout.strip().split('\n')[0]
                    info['bwrap_version'] = version_line
                    
            except:
                pass
        
        features = ['user_namespaces', 'network_isolation', 'filesystem_isolation', 'process_isolation']
        info['supported_features'] = features
        
        return info
    
    def cleanup_processes(self):
        for process_id, process in list(self._running_processes.items()):
            try:
                if process.poll() is None:
                    process.terminate()
                    try:
                        process.wait(timeout=5)
                    except subprocess.TimeoutExpired:
                        process.kill()
                        process.wait()
                
                del self._running_processes[process_id]
            except:
                pass
    
    def get_isolation_recommendations(self, command: List[str]) -> SandboxOptions:
        if not command:
            return SandboxOptions()
        
        cmd_name = os.path.basename(command[0]).lower()
        
        if cmd_name in ['makepkg', 'paru', 'yay']:
            return SandboxOptions(
                isolation_level=IsolationLevel.MEDIUM,
                allow_network=True,
                allow_home=False,
                ro_bind_paths=[
                    ('/var/lib/pacman', '/var/lib/pacman'),
                    ('/var/cache/pacman', '/var/cache/pacman'),
                    ('/etc/pacman.conf', '/etc/pacman.conf'),
                    ('/etc/pacman.d', '/etc/pacman.d')
                ]
            )
        
        elif cmd_name in ['git', 'wget', 'curl']:
            return SandboxOptions(
                isolation_level=IsolationLevel.MEDIUM,
                allow_network=True,
                allow_home=False
            )
        
        elif cmd_name in ['gcc', 'g++', 'make', 'cmake']:
            return SandboxOptions(
                isolation_level=IsolationLevel.MINIMAL,
                allow_network=False,
                allow_home=False
            )
        
        else:
            return SandboxOptions(
                isolation_level=IsolationLevel.MEDIUM,
                allow_network=False,
                allow_home=False
            )

File: src/gear/sandboxing.rs
use std::collections::HashMap;
use std::path::Path;
use std::process::{Command, Stdio};
// use std::io::{BufRead, BufReader};
use std::time::Instant;
use which::which;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum IsolationLevel {
    Strict,
    Medium,
    Minimal,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SandboxStatus {
    Ready,
    Running,
    Completed,
    Failed,
    Timeout,
}

#[derive(Debug, Clone)]
pub struct SandboxOptions {
    pub isolation_level: IsolationLevel,
    pub allow_network: bool,
    pub allow_home: bool,
    pub allow_x11: bool,
    pub allow_dbus: bool,
    pub working_dir: Option<String>,
    pub bind_paths: Vec<(String, String, Option<String>)>,
    pub env_vars: HashMap<String, String>,
    pub capabilities: Vec<String>,
    pub uid: Option<u32>,
    pub gid: Option<u32>,
    pub hostname: Option<String>,
    pub tmpfs_paths: Vec<String>,
    pub ro_bind_paths: Vec<(String, String)>,
    pub dev_bind_paths: Vec<(String, String)>,
}

impl Default for SandboxOptions {
    fn default() -> Self {
        Self {
            isolation_level: IsolationLevel::Medium,
            allow_network: false,
            allow_home: false,
            allow_x11: false,
            allow_dbus: false,
            working_dir: None,
            bind_paths: Vec::new(),
            env_vars: HashMap::new(),
            capabilities: Vec::new(),
            uid: None,
            gid: None,
            hostname: None,
            tmpfs_paths: Vec::new(),
            ro_bind_paths: Vec::new(),
            dev_bind_paths: Vec::new(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct SandboxResult {
    pub return_code: i32,
    pub stdout: String,
    pub stderr: String,
    pub status: SandboxStatus,
    pub execution_time: f64,
    pub warnings: Vec<String>,
    pub errors: Vec<String>,
}

#[derive(Debug)]
pub struct SandboxManager {
    // In a real app, you might want to track processes, but for now we'll keep it simple.
}

impl SandboxManager {
    pub fn new() -> Result<Self, String> {
        if which("bwrap").is_err() {
            return Err("bubblewrap (bwrap) is required but not found".to_string());
        }
        Ok(Self {})
    }

    pub fn run_sandboxed_command(
        &self,
        command: &[String],
        options: &SandboxOptions,
        _timeout: Option<u64>,
    ) -> SandboxResult {
        let start_time = Instant::now();
        let bwrap_args = self._build_bwrap_command(command, options);

        let mut process_cmd = Command::new(&bwrap_args[0]);
        process_cmd.args(&bwrap_args[1..])
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        if let Some(ref dir) = options.working_dir {
            process_cmd.current_dir(dir);
        }

        let mut child = match process_cmd.spawn() {
            Ok(c) => c,
            Err(e) => return SandboxResult {
                return_code: -1,
                stdout: String::new(),
                stderr: e.to_string(),
                status: SandboxStatus::Failed,
                execution_time: start_time.elapsed().as_secs_f64(),
                warnings: Vec::new(),
                errors: vec![format!("Failed to spawn process: {}", e)],
            },
        };

        let result = if let Some(timeout_secs) = _timeout {
            let timeout_dur = std::time::Duration::from_secs(timeout_secs);
            let mut status = SandboxStatus::Running;
            let mut output_res = Option::None;

            while start_time.elapsed() < timeout_dur {
                match child.try_wait() {
                    Ok(Some(s)) => {
                        status = if s.success() { SandboxStatus::Completed } else { SandboxStatus::Failed };
                        output_res = Some(s);
                        break;
                    }
                    Ok(Option::None) => {
                        std::thread::sleep(std::time::Duration::from_millis(100));
                    }
                    Err(_e) => {
                        status = SandboxStatus::Failed;
                        break;
                    }
                }
            }

            if output_res.is_none() && status == SandboxStatus::Running {
                let _ = child.kill();
                status = SandboxStatus::Timeout;
            }
            
            // Collect output after wait/kill
            let output = child.wait_with_output().ok();
            (status, output)
        } else {
            let output = child.wait_with_output().ok();
            let status = match &output {
                Some(o) => if o.status.success() { SandboxStatus::Completed } else { SandboxStatus::Failed },
                Option::None => SandboxStatus::Failed,
            };
            (status, output)
        };

        let (status, output) = result;
        let stdout_str = output.as_ref().map(|o| String::from_utf8_lossy(&o.stdout).to_string()).unwrap_or_default();
        let stderr_str = output.as_ref().map(|o| String::from_utf8_lossy(&o.stderr).to_string()).unwrap_or_default();
        let return_code = output.as_ref().and_then(|o| o.status.code()).unwrap_or(-1);

        SandboxResult {
            return_code,
            stdout: stdout_str.clone(),
            stderr: stderr_str.clone(),
            status,
            execution_time: start_time.elapsed().as_secs_f64(),
            warnings: self._extract_warnings(&stdout_str, &stderr_str),
            errors: self._extract_errors(&stderr_str),
        }
    }

    fn _build_bwrap_command(&self, command: &[String], options: &SandboxOptions) -> Vec<String> {
        let mut args = vec!["bwrap".to_string(), "--unshare-all".to_string(), "--die-with-parent".to_string()];

        match options.isolation_level {
            IsolationLevel::Strict => {
                args.extend(vec![
                    "--unshare-user".to_string(),
                    "--unshare-ipc".to_string(),
                    "--unshare-pid".to_string(),
                    "--unshare-cgroup".to_string(),
                    "--no-new-privileges".to_string(),
                    "--cap-drop".to_string(), "ALL".to_string(),
                ]);
                for cap in &options.capabilities {
                    args.extend(vec!["--cap-add".to_string(), cap.clone()]);
                }
            }
            IsolationLevel::Medium => {
                args.extend(vec![
                    "--unshare-user".to_string(),
                    "--unshare-ipc".to_string(),
                    "--unshare-pid".to_string(),
                    "--unshare-net".to_string(),
                ]);
            }
            IsolationLevel::Minimal => {}
        }

        args.extend(vec!["--dev".to_string(), "/dev".to_string()]);
        args.extend(vec!["--proc".to_string(), "/proc".to_string()]);
        args.extend(vec!["--tmpfs".to_string(), "/tmp".to_string()]);
        args.extend(vec!["--tmpfs".to_string(), "/var/tmp".to_string()]);

        for tmpfs in &options.tmpfs_paths {
            args.extend(vec!["--tmpfs".to_string(), tmpfs.clone()]);
        }

        let essential_ro_binds = ["/usr", "/bin", "/lib", "/lib64", "/sbin", "/etc"];
        for path in essential_ro_binds {
            if Path::new(path).exists() {
                args.extend(vec!["--ro-bind".to_string(), path.to_string(), path.to_string()]);
            }
        }

        for (host, container) in &options.ro_bind_paths {
            if Path::new(host).exists() {
                args.extend(vec!["--ro-bind".to_string(), host.clone(), container.clone()]);
            }
        }

        for (host, container, flags) in &options.bind_paths {
            if Path::new(host).exists() {
                let bind_type = flags.as_deref().unwrap_or("--bind");
                args.extend(vec![bind_type.to_string(), host.clone(), container.clone()]);
            }
        }

        if options.allow_network {
            args.push("--share-net".to_string());
        } else {
            args.push("--unshare-net".to_string());
        }

        if options.allow_home {
            if let Some(home) = dirs::home_dir() {
                let h = home.to_string_lossy().to_string();
                args.extend(vec!["--bind".to_string(), h.clone(), h]);
            }
        } else {
            args.extend(vec!["--tmpfs".to_string(), "/home".to_string()]);
            args.extend(vec!["--mkdir".to_string(), "/home/sandbox".to_string()]);
            args.extend(vec!["--setenv".to_string(), "HOME".to_string(), "/home/sandbox".to_string()]);
        }

        if let Some(ref dir) = options.working_dir {
            args.extend(vec!["--bind".to_string(), dir.clone(), dir.clone()]);
            args.extend(vec!["--chdir".to_string(), dir.clone()]);
        } else {
            args.extend(vec!["--chdir".to_string(), "/tmp".to_string()]);
        }

        if let Some(uid) = options.uid {
            args.extend(vec!["--uid".to_string(), uid.to_string()]);
        }
        if let Some(gid) = options.gid {
            args.extend(vec!["--gid".to_string(), gid.to_string()]);
        }
        if let Some(ref hostname) = options.hostname {
            args.extend(vec!["--hostname".to_string(), hostname.clone()]);
        }

        args.extend(vec!["--setenv".to_string(), "PATH".to_string(), "/usr/bin:/bin:/usr/sbin:/sbin".to_string()]);

        for (key, val) in &options.env_vars {
            args.extend(vec!["--setenv".to_string(), key.clone(), val.clone()]);
        }

        args.push("--".to_string());
        args.extend(command.to_vec());

        args
    }

    fn _extract_warnings(&self, stdout: &str, stderr: &str) -> Vec<String> {
        let mut warnings = Vec::new();
        let combined = format!("{}{}", stdout, stderr);
        let patterns = ["warning:", "caution:", "deprecated", "bwrap:"];
        for line in combined.lines() {
            let line_lower = line.to_lowercase();
            if patterns.iter().any(|p| line_lower.contains(p)) {
                warnings.push(line.trim().to_string());
            }
        }
        warnings
    }

    fn _extract_errors(&self, stderr: &str) -> Vec<String> {
        let mut errors = Vec::new();
        let patterns = ["error:", "failed:", "cannot", "unable to", "not found"];
        for line in stderr.lines() {
            let line_lower = line.to_lowercase();
            if patterns.iter().any(|p| line_lower.contains(p)) {
                errors.push(line.trim().to_string());
            }
        }
        errors
    }
}
